### ISSUE TITLE
Max pooling issue when stride is 1

### ISSUE THREAD

**Issue by ersheng-ai** [Tianxiaomo/pytorch-YOLOv4#62]
In the original DarkNet model, the there are 3 max pooling layers of kernel size:  5, 9, 13, and stride is 1 for all of them.
But the solution in darknet2pytorch is confusing.
I don't know why kernel size is set to 2 when stride is 1 because it would not be mathematical identical.

```python
                padding = 0
                if 'pad' in block.keys() and int(block['pad']) == 1:
                    padding = int((pool_size - 1) / 2)
                if stride > 1:
                    model = nn.MaxPool2d(pool_size, stride, padding=padding)
                else:
                    model = MaxPoolStride1()
```
and
```python
class MaxPoolStride1(nn.Module):
    def __init__(self, size=2):
        super(MaxPoolStride1, self).__init__()
        self.size = size
        if (self.size - 1) % 2 == 0:
            self.padding1 = (self.size - 1) // 2
            self.padding2 = self.padding1
        else:
            self.padding1 = (self.size - 1) // 2
            self.padding2 = self.padding1 + 1

    def forward(self, x):
        x = F.max_pool2d(F.pad(x, (self.padding1, self.padding2, self.padding1, self.padding2), mode='replicate'),
                         self.size, stride=1)
        return x
```

**Comment by dhiyu** [comment]
I think it may be `model = MaxPoolStride1(pool_size)`. But any one knows what is different between  `stride > 1` and  `stride == 1`?

**Comment by ersheng-ai** [comment]
> I think it may be `model = MaxPoolStride1(pool_size)`. But any one knows what is different between `stride > 1` and `stride == 1`?

You can simply do this to temporarily bypass this issue:
```python
if pool_size == 5:
    model = nn.MaxPool2d(5,  1,  padding=2)
elif pool_size == 9:
    model = nn.MaxPool2d(9,  1,  padding=4)
else: # pool_size == 13
    model = nn.MaxPool2d(13,  1,  padding=6)
```

**Comment by Tianxiaomo** [comment]
This is result of the difference between darknet and pytorch's implementation of padding.

darknet parser.c
```
maxpool_layer parse_maxpool(list *options, size_params params)
{
...
    int size = option_find_int(options, "size",stride);
    int padding = option_find_int_quiet(options, "padding", size-1);
...
   maxpool_layer layer = make_maxpool_layer(batch, h, w, c, size, stride_x, stride_y, padding, maxpool_depth, out_channels, antialiasing, avgpool, params.train);
    return layer;
}
```
darknet maxpool_layer.c
```
maxpool_layer make_maxpool_layer(int batch, int h, int w, int c, int size, int stride_x, int stride_y, int padding, int maxpool_depth, int out_channels, int antialiasing, int avgpool, int train)
{
...
    if (maxpool_depth) {
        l.out_c = out_channels;
        l.out_w = l.w;
        l.out_h = l.h;
    }
    else {
        l.out_w = (w + padding - size) / stride_x + 1;
        l.out_h = (h + padding - size) / stride_y + 1;
        l.out_c = c;
    }
...
}
```
It can be found that the darknet's pad parameter is set to '1' and the padding is 'SAME' in tensorflow is the same.
but in pytorch,
padding: implicit zero padding to be added on both sides.

So I do n’t have a beautiful way to deal with it.:sweat:

**Comment by ersheng-ai** [comment]
Make the solution perfect and universal is difficult but the minimum requirement is that you make YOLOv4/v3 correct.
So, take it easy.
